[MRG+1] Uncontroversial fixes from estimator tags branch

amueller opened the pull request:
These are some of the simple fixes from #8022 which should be fairly uncontroversial and easy to review.
I hope that makes the review of #8022 easier and allows that PR to focus on the API.

amueller committed:
some bug fixes.

amueller commented on doc/whats_new.rst:
```diff
@@ -127,6 +127,14 @@ Bug fixes
+   - Fixes to the input validation in :class:`sklearn.covariance.EllipticEnvelope` by
+     `Andreas Müller`_.
+
+   - Fix shape output shape of :class:`sklearn.decomposition.DictionaryLearning` transform
```
this still needs a test, I think...

amueller committed:
minor fixes to whatsnew

amueller commented on doc/whats_new.rst:
```diff
@@ -136,6 +147,29 @@ API changes summary
+
+   - Gradient boosting base models are not longer estimators. By `Andreas Müller`_.
+
+   - :class:`feature_extraction.text.TfidfTransformer` now supports numpy
```
this needs tests, I guess

amueller committed:
typo in whatsnew

amueller committed:
add test for n_components = 1 transform in dict learning

amueller commented on sklearn/neighbors/approximate.py:
```diff
@@ -93,7 +93,7 @@ class GaussianRandomProjectionHash(ProjectionToHashMixin,
     """Use GaussianRandomProjection to produce a cosine LSH fingerprint"""
     def __init__(self,
-                 n_components=8,
+                 n_components=32,
```
This class is never instantiated with default parameters and the code doesn't run with n_compenents != 32

jnothman replied:
ha!

agramfort replied:
is the error captured if not 32? should it be a parameter if only one  param value works?

amueller replied:
I told @ogrisel to look into it ;)

amueller commented on sklearn/feature_extraction/text.py:
```diff
@@ -1056,18 +1056,19 @@ def transform(self, X, copy=True):
-        else:
-            # convert counts or binary occurrences to floats
-            X = sp.csr_matrix(X, dtype=np.float64, copy=copy)
+        X = check_array(X, accept_sparse=["csr"], copy=copy,
```
I'm not actually somewhat uncertain of whether this is the right approach. In ``fit`` we always convert to sparse. Maybe we should be consistent? I'm not sure.

jnothman replied:
I think the principle is that for `df` to be meaningful, `X` needs to be fundamentally sparse (regardless of data structure). I can see why you consider this broken, but I'm tempted to say don't fix. You're creating a backwards-incompatibility.

amueller replied:
reverted

amueller committed:
feature extraction doc fix

TomDLT commented on doc/whats_new.rst:
```diff
@@ -136,6 +147,29 @@ API changes summary
+
+   - :class:`feature_extraction.text.TfidfTransformer` now supports numpy
+     arrays as inputs, and produces numpy arrays for list inputs and numpy
+     array inputs. By `Andreas Müller_`.
```
swap _ and `

jnothman commented on sklearn/covariance/outlier_detection.py:
```diff
@@ -176,3 +177,29 @@ def fit(self, X, y=None):
             self.dist_, 100. * (1. - self.contamination))
         return self
+
+    def score(self, X, y, sample_weight=None):
```
Why is this not part of `OutlierDetectionMixin`?

amueller replied:
moved it there.

jnothman commented on sklearn/dummy.py:
```diff
@@ -117,6 +117,9 @@ def fit(self, X, y, sample_weight=None):
 
         self.sparse_output_ = sp.issparse(y)
 
+        X = check_array(X, accept_sparse=['csr', 'csc', 'coo'])
```
I think we need to be liberal wrt `X`: `DummyClassifier` will often substitute for a pipeline / grid search, where `X` may be a list of strings, a dataframe, a list of open files, or other mess. Don't demand too much of it. Here you ensure it is 2d and numeric and of a particular format, but why?

amueller replied:
removed.

jnothman commented on doc/whats_new.rst:
````diff
@@ -136,6 +147,29 @@ API changes summary
+
+   - :class:`feature_selection.SelectFromModel` now validates the ``threshold``
+     parameter and sets the ``threshold_`` attribute during the call to
+     ``fit``, and no longer during the call to ``transform```, by `Andreas
````
I think it's still being set in `transform` now, no?

jnothman commented on sklearn/naive_bayes.py:
```diff
@@ -480,13 +480,13 @@ def partial_fit(self, X, y, classes=None, sample_weight=None):
             Target values.
 
-        classes : array-like, shape = [n_classes], optional (default=None)
+        classes : array-like, shape = [n_classes], (default=None)
```
no comma before parentheses?

jnothman commented on sklearn/naive_bayes.py:
```diff
@@ -480,13 +480,13 @@ def partial_fit(self, X, y, classes=None, sample_weight=None):
             in subsequent calls.
 
-        sample_weight : array-like, shape = [n_samples], optional (default=None)
+        sample_weight : array-like, shape = [n_samples], (default=None)
```
no comma before parentheses?

jnothman reviewed (commented):
otherwise LGTM, I think.

Thanks! But I can't say it's so easy to review a heterogeneous patch like this, and I wish you'd pulled the entirely cosmetic things out separately.

amueller committed:
fix broken test

GaelVaroquaux commented on sklearn/ensemble/gradient_boosting.py:
```diff
@@ -64,7 +64,7 @@
 
 
-class QuantileEstimator(BaseEstimator):
+class QuantileEstimator(object):
```
Why this change? (there is probably a good reason, just asking)

amueller replied:
These are not scikit-learn estimators, they don't fulfill the sklearn estimator API and the inheritance doesn't provide any functionality. So I think having them inherit is rather confusing.
The more direct reason is that I don't want these to be discovered by the common tests, as they are not sklearn estimators and will definitely fail the tests.

jnothman replied:
They are also not in a position for get_params/set_params to be used.

GaelVaroquaux commented:
Overall +1 for merge, but I added a small comment. It also seems that there are unanswered comments by @jnothman

amueller committed:
revert aggressive input validation changes

amueller committed:
in SelectFromModel, don't store threshold_ in transform. If we called "fit", use estimates from last "fit".

amueller commented on doc/whats_new.rst:
```diff
@@ -267,6 +279,25 @@ API changes summary
+   - Gradient boosting base models are not longer estimators. By `Andreas Müller`_.
+
+   - :class:`feature_selection.SelectFromModel` now validates the ``threshold``
+     parameter and sets the ``threshold_`` attribute during the call to
```
Now actually doesn't set it any more during transform. This is a backward incompatible change, but could be considered a bug in the previous implementation? I tried to make it clearer what happens now.

amueller committed:
Merge branch 'master' into tags_backup

amueller committed:
move score from EllipticEnvelope to OutlierDetectionMixin

amueller committed:
revert changes to Tfidf documentation

amueller committed:
remove dummy input validation from whatsnew

amueller committed:
fix text feature tests

amueller committed:
rewrite from_model threshold again...

amueller commented:
Ok so I reworked the ``SelectFromModel`` again. The point of ``self.threshold_`` is to provide a threshold to the user, it's not really used internally (before or after this PR). However, before the PR, if the user re-assigned ``threshold`` - a use-case that we explicitly support and test - then ``self.threshold_`` is outdated until ``transform`` is called. Even calling ``fit`` would not update it!

The PR now makes ``threshold_`` a property so it doesn't go stale. Computing it is quick so it's not a big deal. We can think about deprecating the property and making it a method.

jnothman commented:
I've not looked, but sounds good

On 16 May 2017 8:37 am, "Andreas Mueller" <notifications@github.com> wrote:

> Ok so I reworked the SelectFromModel again. The point of self.threshold_
> is to provide a threshold to the user, it's not really used internally
> (before or after this PR). However, before the PR, if the user re-assigned
> threshold - a use-case that we explicitly support and test - then
> self.threshold_ is outdated until transform is called. Even calling fit
> would not update it!
>
> The PR now makes threshold_ a property so it doesn't go stale. Computing
> it is quick so it's not a big deal. We can think about deprecating the
> property and making it a method.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/scikit-learn/scikit-learn/pull/8086#issuecomment-301624869>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AAEz63bjbnRiEjugvUTvBY-BlvfyDM6Tks5r6NPBgaJpZM4LRPhH>
> .
>

amueller commented:
Can anyone explain to me what codecov is doing? I'm looking at https://codecov.io/gh/scikit-learn/scikit-learn/compare/master...tags_backup/changes but the changes seem to be in the docstrings?!

amueller committed:
remove stray condition

jnothman commented on sklearn/feature_extraction/text.py:
```diff
@@ -1088,10 +1088,9 @@ def transform(self, X, copy=True):
-        else:
-            return None
+        # if _idf_diag is not set, this will raise an attribute error,
+        # which means hasatt(self, "idf_") is False
```
*hasattr

jnothman replied:
Also, test this?

amueller replied:
It is tested :)

jnothman commented on sklearn/feature_selection/from_model.py:
```diff
@@ -169,6 +169,12 @@ def fit(self, X, y=None, **fit_params):
 
+    @property
+    def threshold_(self):
+        scores = _get_feature_importances(self.estimator_, self.norm_order)
```
This can be expensive. So where a user used to be able to use `sfm.threshold_` with impunity, now it may take some time.

amueller replied:
When is this expensive? I guess if you have a many classes and many features the reduction to the vector is expensive?

Before, when they used ``sfm.threshold_`` it might have been wrong ;)
We could store the scores in fit and / or transform. If people re-assign ``estimator`` that would go stale, though. So we could support re-assigning ``threshold`` but not ``estimator``?

amueller replied:
If we cache scores, then the property might still require recomputing the mean or std. But we can't avoid that if we want ``threshold_`` to be up-to-date with ``threshold`` unless we overwrite ``setattr`` which I'm not sure is a good idea ;)

amueller replied:
no matter if they are used for computing the threshold or not. I could make the computation dependent on whether ``self.threshold`` is a string type.

vene replied:
I agree, the cost of evaluating the column-wise norm and the mean/median shouldn't be so significant. Can't think of use cases where this would be called in a tight loop.

GaelVaroquaux replied:
What is the logic for making this a property, compared to setting it in fit?

amueller replied:
we explicitly allow people to change a threshold after fitting the estimator. There's a test for that.

jnothman commented on sklearn/feature_selection/from_model.py:
```diff
@@ -136,12 +136,12 @@ def _get_support_mask(self):
-        self.threshold_ = _calculate_threshold(estimator, scores,
-                                               self.threshold)
-        return scores >= self.threshold_
+        threshold = _calculate_threshold(self.estimator, scores,
```
should use `estimator`, not `self.estimator`

jnothman replied:
Needs a test

amueller replied:
Actually, doesn't matter mostly. ``_calculate_threshold`` only checks whether ``estimator.penalty=='l1'`` in some cases. I *guess* you could come up with a case where that's true for ``estimator`` but not ``estimator_`` or vice versa, but changing ``penalty`` during fit is actually violating API constraints, right? (because no trailing underscore).
I changed it, because it's more obvious.

What did you want tested?

jnothman commented:
Yes, some of that codecov output is weird.

amueller committed:
fix self.estimator -> estimator, slightly more interesting test

amueller committed:
typo in comment

amueller committed:
Merge branch 'master' into tags_backup

amueller commented:
can I has review? ;)

GaelVaroquaux commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -282,9 +282,9 @@ def sparse_encode(X, dictionary, gram=None, cov=None, algorithm='lasso_lars',
+        # consistent with the case n_jobs > 1
         if code.ndim == 1:
-            code = code[np.newaxis, :]
+            code = code[:, np.newaxis]
```
I guess that this is the controversial part of the PR.

Do we have any estimation on how much user code we will break? Maybe we should make this change more visible in the whats_new

amueller replied:
You mean fix? ;) This is a bugfix, but I'm happy to make it more visible in the whatsnew. It's been a while since I wrote this, I need to look up the exact edge cases this fixes.

The controversial parts where supposed to be the complete refactoring of all the common tests and the introduction of  estimator tags. This PR fixes all the bugs that better testing finds.

GaelVaroquaux replied:
OK, your logic is that the return dimensions varied whether n_jobs=1 or not, hence this is a clear-cut bug?

I am worried that some users might rely on this behavior. Upgrading scikit-learn might break there code. Do you think it's unlikely? I don't know, I don't personally use this part of scikit-learn.

This is why I was suggesting a more prominent entry in whats_new

amueller replied:
It's a clear-cut bug fix for its use in DictionaryLearning, which was broken with ``n_components=1``. I need to double check how n_jobs affects this and what this means for people that directly use SparseEncode

amueller replied:
(one direction of code needs to be n_features so their could only be ambiguity when n_features=1. Both n_components=1 and n_features=1 seems odd, but I'll add some tests that make sure both work as expected and see if this has any behavior change.

amueller replied:
n_components=1 is still broken with n_jobs > 1

amueller replied:
Sorry my second to last comment doesn't make sense, I was confused about code vs dictionary.

amueller commented:
I can also split this up into one PR per bugfix? seems like the changes are not that simple after all... In particular without regression tests. The point is kind of that adding the estimator tags will find all these bugs, but reviewing without regression tests might be too tricky.

GaelVaroquaux commented:
WRT splitting this PR in smaller one: it is indeed a though review, but I think that we need to make sure that you don't loose time. I think that we can review and merge this guy in the next few days.

amueller commented:
@GaelVaroquaux ok cool. I was hoping that we can discuss the main PR #8022 soon, as this will fix a whole bunch of issues with the testing.

amueller commented:
@vene this one is the first step to more contrib friendly tests ;)

amueller commented on sklearn/decomposition/tests/test_dict_learning.py:
```diff
@@ -28,7 +28,12 @@
-    assert_true(dico.components_.shape == (n_components, n_features))
+    assert_equal(dico.components_.shape, (n_components, n_features))
+
+    n_components = 1
```
@GaelVaroquaux this is the regression test for the SparseEncode change. I can add a more direct test on SparseEncode, though.

GaelVaroquaux replied:
No, I think that this is good.

amueller replied:
Added more test now, one testing the core issue in ``sparse_encode``, the other one the resulting bug in ``DictonaryLearning``

amueller committed:
Fix issues in SparseEncoder, add tests.
more explicit explanation of SparseEncoder change, add issue numbers to whatsnew

amueller committed:
Merge branch 'master' into tags_backup

# Conflicts:
#	.travis.yml

amueller committed:
Merge branch 'tags_backup' of github.com:amueller/scikit-learn into tags_backup

amueller commented:
Ok I completely rewrote the fix for DictionaryLearning and added tests for ``sparse_encode`` directly. Previously n_jobs>1 code-paths gave ValueError and sparse encode and dictionary learning returned nonsensical results in several cases.

GaelVaroquaux commented:
> Previously n_jobs>1 code-paths gave ValueError and sparse encode and
> dictionary learning returned nonsensical results in several cases.

OK, sorry, I had not realized this.

amueller commented:
@GaelVaroquaux neither did I. I just delved into it with @vene ;)

vene commented on sklearn/feature_selection/tests/test_from_model.py:
```diff
@@ -171,10 +176,10 @@ def test_threshold_string():
 
     # Set a higher threshold to filter out more features.
-    model.threshold = 1.0
+    model.threshold = "1.0 * mean"
```
is this just for covering the expression parsing? The tests should not fail with the old values, right?

amueller replied:
It was also to make sure that we're recomputing ``self.threshold_``

vene commented:
My only remaining controversy in this PR is whether we want to change the `GaussianRandomProjections` default n_components like that, or fix that in a separate PR. (But I don't want to delay this.)

amueller commented:
Is @ogrisel around? I think he did the GaussianRandomProjection. It's only used internally. I guess we should open an issue for that?

vene commented:
issue #8029 also mentions that these classes are not the greatest. (btw this is `GaussianRandomProjectionHash`, sorry for the typo.)

I'll comment under #8029 with the weird behaviour we are finding.

agramfort commented on sklearn/covariance/outlier_detection.py:
```diff
@@ -100,8 +101,34 @@ def predict(self, X):
+        X : array-like, shape = (n_samples, n_features)
+            Test samples.
+
+        y : array-like, shape = (n_samples) or (n_samples, n_outputs)
```
shape = (n_samples,) or (n_samples, n_outputs)

agramfort commented on sklearn/covariance/outlier_detection.py:
```diff
@@ -100,8 +101,34 @@ def predict(self, X):
+        y : array-like, shape = (n_samples) or (n_samples, n_outputs)
+            True labels for X.
+
+        sample_weight : array-like, shape = [n_samples], optional
```
array-like, shape = (n_samples,), optional

agramfort commented on sklearn/naive_bayes.py:
```diff
@@ -484,13 +484,13 @@ def partial_fit(self, X, y, classes=None, sample_weight=None):
             Target values.
 
-        classes : array-like, shape = [n_classes], optional (default=None)
+        classes : array-like, shape = [n_classes] (default=None)
```
I see that we are not consistent here.

I think we should use tuples for shapes in doctrings so:

array-like, shape = (n_classes,) (default=None)

amueller replied:
well I could change that but that would really increase the size of the pull request, even if I just do it for naive_bayes.py

TomDLT commented on doc/whats_new.rst:
````diff
@@ -336,6 +352,20 @@ API changes summary
+     ``fit``, and no longer during the call to ``transform```, by `Andreas
+     Müller`_.
+
+   - :class:`features_selection.SelectFromModel` now has a ``partial_fit``
````
features -> feature

agramfort commented:
that's it for me.

TomDLT commented on doc/whats_new.rst:
```diff
@@ -249,6 +251,20 @@ Bug fixes
+     :issue:`8086` by `Andreas Müller`_.
+
+   - Fix output shape and bugs with n_jobs > 1 in  
+     :class:`sklearn.decomposition.SparseEncoder` transform and :func:`sklarn.decomposition.sparse_encode`
```
SparseEncoder -> SparseCoder

vene commented on sklearn/decomposition/dict_learning.py:
```diff
@@ -94,6 +94,11 @@ def _sparse_encode(X, dictionary, gram, cov=None, algorithm='lasso_lars',
+    if dictionary.shape[1] != X.shape[1]:
+        raise ValueError("Dictionary and X have different numbers of features:"
+                         "dictionary.shape: {} X.shape{}".format(
+                             dictionary.shape, X.shape))
```
could use `check_consistent_length(X.T, dictionary.T)`

You could argue that's less clear though.

amueller replied:
It'll say "different number of samples" in the error, right?

TomDLT commented on doc/whats_new.rst:
```diff
@@ -213,7 +213,8 @@ Bug fixes
    - Fixed a bug where :class:`sklearn.linear_model.LassoLars` does not give
      the same result as the LassoLars implementation available
-     in R (lars library). :issue:`7849` by :user:`Jair Montoya Martinez <jmontoyam>`
+     in R (lars library). :issue:`7849` by `Jair Montoya Martinez`_
```
There is no link for Jair Montoya Martinez.
Revert this one?

agramfort replied:
use this url : https://github.com/jmontoyam

amueller replied:
thanks, I guess that got lost in a merge.

TomDLT replied:
> use this url : https://github.com/jmontoyam

Then revert it to 
```
:user:`Jair Montoya Martinez <jmontoyam>`
```

TomDLT commented on doc/whats_new.rst:
```diff
@@ -223,9 +224,10 @@ Bug fixes
-     being subtracted from the centroids. :issue:`7872` by `Josh Karnofsky <https://github.com/jkarno>`_.
+   - Fix a bug regarding fitting :class:`sklearn.cluster.KMeans` with a sparse
+     array X and initial centroids, where X's means were unnecessarily being
+     subtracted from the centroids. :issue:`7872` by `Josh Karnofsky
```
:user:`Josh Karnofsky <jkarno>`

amueller committed:
minor fixes in whats_new.rst

amueller committed:
slightly more consistency with tuples for shapes

agramfort commented:
good to go on my end

vene commented on doc/whats_new.rst:
```diff
@@ -336,6 +351,20 @@ API changes summary
      :issue:`2879` by :user:`Stephen Hoover <stephen-hoover>`.
 
+
+   - Gradient boosting base models are not longer estimators. By `Andreas Müller`_.
```
not longer -> no longer

vene commented:
I was intrigued by the bugfix of `self.n_outputs_ == 1` and found that there is another instance of this conditional that is missing the `and not self.output_2d_` guard at https://github.com/scikit-learn/scikit-learn/pull/8086/files#diff-a5d4091095f8d33282dd04afcfa8d2d5R322, is that a bug that was not caught by this round of common tests?

amueller committed:
not longer typo

vene commented:
Since we plan to deprecate the GaussianRandomProjectionHash class in this cycle, maybe we should simply skip it from the common tests rather than change its default n_components. I really doubt that class is seeing any external use, and internally it is only ever called with a fixed n_components.

Other than the very minor `check_consistent_length` question in dict learning this PR looks good to me.

amueller commented:
@GaelVaroquaux do you wanna have a final look or merge?

amueller commented:
@vene merge?
